### PR TITLE
CI Improve pre-commit setup and run it on the CI

### PR DESCRIPTION
**Pull request by jjerphan**
This:
 - tune black correctly
 - use `isort` to sort import canonically
 - add a GitHub workflow to run format checks

Potentially we could have a single workflow with two jobs (one for those checks and another for the CI). What do you think?

### CODE REVIEW

File: setup.py
Comment (ogrisel) [inline setup.py:40]: Thanks for fixing this! I could not understand what was wrong myself!

File: sklearn_numba_dpex/kmeans/kernels.py
Comment (ogrisel) [inline sklearn_numba_dpex/kmeans/kernels.py:82]: We could keep this comment consistently aligned with the others here.

```suggestion
        new_centroids_t_private_copies,    # OUT            (n_private_copies, n_features, n_clusters)  # noqa
        cluster_sizes_private_copies,      # OUT            (n_private_copies, n_clusters)  # noqa
```

File: setup.py
Comment (jjerphan) [reply setup.py:40]: An extra character and the world is broken. :upside_down_face:

I find the inconsistency in the naming of `install_requires` and `extra_require` weird and error-prone as shown here.

File: sklearn_numba_dpex/kmeans/kernels.py
Comment (ogrisel) [inline sklearn_numba_dpex/kmeans/kernels.py:84]: Question: what does this stand for?

**Review by ogrisel — APPROVED** [review]
Thanks for setting this up. +1 for me if CI is green.

**Comment by ogrisel** [comment]
> Potentially we could have a single workflow with two jobs (one for those checks and another for the CI). What do you think?

No strong opinion. I like having 2 different lines in the status summary of the PR on github.

File: sklearn_numba_dpex/kmeans/kernels.py
Comment (jjerphan) [reply sklearn_numba_dpex/kmeans/kernels.py:82]: Rah, there is some hardness in having black ignore this block in the signature of functions.

**Comment by jjerphan** [comment]
> No strong opinion. I like having 2 different lines in the status summary of the PR on github.

Good to know, and so do I, hence the current setup.

**Comment by ogrisel** [comment]
```
/opt/venv/lib/python3.9/site-packages/setuptools/_distutils/dist.py:262: UserWarning: Unknown distribution option: 'extra_require'
```

Hum...

```
error: invalid command 'bdist_wheel'
```

is the `wheel` package needed as an explicit dependency?

**Comment by jjerphan** [comment]
Hum, there is something inconsistent with the setup of the CI on `main`.

File: sklearn_numba_dpex/kmeans/kernels.py
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/kernels.py:82]: `fmt: off` and `fmt: on` what aimed at having black ignore the block. sad that it doesn't work :cry:

File: sklearn_numba_dpex/kmeans/kernels.py
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/kernels.py:84]: it was supposed to disable black within the block delimited by those tags

**Review by fcharras — APPROVED** [review]

File: setup.py
Comment (fcharras) [inline setup.py:40]:
```suggestion
    extras_require=[dict(benchmark=["scikit-learn-intelex, pandas"])],
```
apparently there's an "s" on extra

**Comment by fcharras** [comment]
Cool initiative very needed, thank you, LGTM when it's green, would be a plus to have black correctly ignores the signature blocks in the kernels .

The current issue seems to come from the file pyproject.toml .

**Comment by fcharras** [comment]
I've looked a bit more into the `pyproject.toml` format and I think that your use of `[build-system]` is misguided, its use is closer to `setup_requires` than `install_requires`. If we want to configure the dependency tree of the package in this file the relevant config blocks are [dependencies/optional-dependencies](https://packaging.python.org/en/latest/specifications/declaring-project-metadata/#dependencies-optional-dependencies).

Also we should not have several sources for the dependencies, it's either `setup.py`, `setup.cfg` or `pyproject.toml`. From the look of it, nowadays the motion is going towards `pyproject.toml` so let's use that exclusively if we can.

**Comment by fcharras** [comment]
Regarding the current `workflow` file, why pass `--no-build-isolation` to `pip` rather than using the default ? one issue was that if not using build isolation pip expects some dependencies (like `wheels`, apparently bdist_wheels are a must have nowadays) to be pre-installed. So we can either explicitly install wheel manually or accept build isolation, since we install only a single package it doesn't make much of a difference. (also, maybe adding wheels as a `setup_requires` would work)

**Comment by fcharras** [comment]
I'll add an additional commit that moves all the setup information to the `pyproject,toml`

**Comment by jjerphan** [comment]
> Regarding the current workflow file, why pass --no-build-isolation to pip rather than using the default ?

Habits from local development. This is indeed not needed here.

**Comment by fcharras** [comment]
So I've pushed more things into the `pyproject.toml`. 

Honestly I think it's OK to have configuration spread into `setup.py`, `pyproject.toml` and `setup.cfg`, Some of those configs seems to still have some specs that are not shared with others.

What is not OK would have to be a same config repeated in different places (and worse, with different values).

Python packaging being what it is.

**Comment by ogrisel** [comment]
Since this is a pure python project (no compiled dependencies) we could probably put everything in `pyproject.toml` with the `flit` packager I think. But this looks already like a net improvement. Let's merge since CI is green.

**Pull request merged**
